### ISSUE TITLE
fix(vlm): validation crashes with FusedLinearCrossEntropy — out.hidden_states[-1] indexes the batch dim when the model returns a single hidden-state tensor

### ISSUE THREAD

**Issue by GITsologun** [NVIDIA-NeMo/Automodel#3206]
## Bug Description

Training a VLM (e.g. Qwen3-VL-30B-A3B MoE) with `FusedLinearCrossEntropy` runs fine
until the **first validation epoch**, where every rank crashes with an
`AssertionError` inside `cut_cross_entropy`:

```
[rank0]:   File ".../nemo_automodel/recipes/vlm/finetune.py", line 1181, in _run_validation_epoch
[rank0]:     local_loss = calculate_loss(
[rank0]:   File ".../nemo_automodel/recipes/vlm/finetune.py", line 485, in calculate_loss
[rank0]:     return loss_fn(**loss_fn_kwargs)
[rank0]:   File ".../nemo_automodel/components/loss/linear_ce.py", line 161, in forward
[rank0]:     loss = linear_cross_entropy(
[rank0]:   File ".../cut_cross_entropy/cce.py", line 172, in cce_linear_cross_entropy
[rank0]:     assert e.size()[0:-1] == targets.size()
[rank0]: AssertionError
```

## Steps to Reproduce

1. Configure a VLM finetune run with:
   ```yaml
   loss_fn:
     _target_: nemo_automodel.components.loss.linear_ce.FusedLinearCrossEntropy
   model:
     _target_: nemo_automodel.NeMoAutoModelForImageTextToText.from_pretrained
     pretrained_model_name_or_path: Qwen/Qwen3-VL-30B-A3B-Instruct
     text_config:
       output_hidden_states: true
   step_scheduler:
     val_every_steps: 100
   validation_dataset: <any validation dataset>
   ```
2. Run training until the first validation is triggered (step 100 with the config above).
3. All ranks raise `AssertionError` in `cut_cross_entropy/cce.py`.

Training itself is unaffected — the crash only happens in `_run_validation_epoch`.

## Root Cause

In `_run_validation_epoch` (`nemo_automodel/recipes/vlm/finetune.py`), hidden states
are extracted with:

```python
hidden_states=out.hidden_states[-1]
if getattr(out, "hidden_states", None) is not None
else None,
```

This assumes HF-style per-layer `hidden_states` (a `tuple[Tensor, ...]`), where `[-1]`
selects the last layer.

However, custom NeMo models return a **single tensor** via
`compute_lm_head_logits` (`nemo_automodel/components/models/common/utils.py`):

```python
return CausalLMOutputWithPast(logits=logits, hidden_states=hidden_out)  # [B, S, H]
```

For a single `[B, S, H]` tensor, `out.hidden_states[-1]` indexes the **last batch
element**, producing `[S, H]`. The fused loss then receives
`e.size()[:-1] == [S]` against `targets.size() == [B, S]`, which fails the
`assert e.size()[0:-1] == targets.size()` check.

The training path does **not** hit this because `_forward_backward_step` already uses
`get_final_hidden_states(out)` from
`nemo_automodel/components/training/model_output_utils.py`, which handles both the
tuple and single-tensor layouts.

## Expected Behavior

Validation extracts the final hidden states the same way training does, so
`FusedLinearCrossEntropy` receives a `[B, S, H]` tensor matching `labels` `[B, S]`.

## Proposed Fix

Align the validation path with the training path in
`nemo_automodel/recipes/vlm/finetune.py::_run_validation_epoch`:

```diff
-                        hidden_states=out.hidden_states[-1]
-                        if getattr(out, "hidden_states", None) is not None
-                        else None,
+                        hidden_states=get_final_hidden_states(out),
```

`get_final_hidden_states` returns the last non-None entry for HF-style tuples
(identical behavior to `[-1]` for HF models) and passes single tensors through
unchanged (fixing custom NeMo models).

The same latent bug exists in two more places and should be fixed as well:
- `nemo_automodel/recipes/vlm/kd.py` (~line 442): `student_out.hidden_states[-1]`
- `nemo_automodel/recipes/llm/kd.py` (~line 652): `student_out.hidden_states[-1]`

## Environment

- Model: Qwen3-VL-30B-A3B-Instruct (custom `qwen3_vl_moe` backend)
- Parallelism: FSDP2, `ep_size=8`, activation checkpointing on
- Loss: `FusedLinearCrossEntropy` (cut-cross-entropy)
- Crash point: first validation epoch (`val_every_steps: 100`)

**Comment by HuiyingLi** [comment]
Thank you @GITsologun , would you mind sending a fix also? Appreciate it!

**Comment by HuiyingLi** [comment]
Would you mind also fixing the kd.py ? many thanks!
